File: src/test/java/com/runninglane/dto/buddy/test/java/cases/bytecode/ByteCodeStrategyTest.java
package com.runninglane.dto.buddy.test.java.cases.bytecode;

import com.runninglane.dto.buddy.javainterop.DtoBuddy;
import com.runninglane.dto.buddy.javainterop.bytecode.ByteCodeStrategy;
import com.runninglane.dto.buddy.javainterop.bytecode.codegen.JavaCodeGenBasedByteCodeStrategy;
import com.runninglane.dto.buddy.javainterop.bytecode.codegen.JavaCodeGenerator;
import com.squareup.javapoet.AnnotationSpec;
import com.squareup.javapoet.TypeSpec;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class ByteCodeStrategyTest {
    public static abstract class TestDto {
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface TestAnnotation {
        String value();
    }

    static class TestByteBuddyByteCodeGenerator extends JavaCodeGenerator {
        @Override
        public @NotNull TypeSpec.Builder defineClass(
            @NotNull Class<?> baseClass,
            List<? extends Class<?>> typeParams,
            @NotNull String packageName,
            @NotNull String className,
            Object dataCollector
        ) {
            return super.defineClass(baseClass, typeParams, packageName, className, dataCollector)
                .addAnnotation(AnnotationSpec.builder(TestAnnotation.class)
                    .addMember("value", "$S", "test123")
                    .build());
        }
        
    }

    @Test
    public void testAddSecondaryConstructor() {
        ByteCodeStrategy byteCodeStrategy = new JavaCodeGenBasedByteCodeStrategy(new TestByteBuddyByteCodeGenerator());
        DtoBuddy dtoBuddy = new DtoBuddy(byteCodeStrategy);
        Class<?> concreteClass = dtoBuddy.implement(TestDto.class);
        TestAnnotation annotation = concreteClass.getAnnotation(TestAnnotation.class);
        assertNotNull(annotation);
        assertEquals("test123", annotation.value());
    }
}

File: src/test/kotlin/com/runninglane/dto/buddy/test/cases/bytecode/ByteCodeStrategyTest.kt
package com.runninglane.dto.buddy.test.cases.bytecode

import com.runninglane.dto.buddy.DtoBuddy
import com.runninglane.dto.buddy.bytecode.bytebuddy.ByteBuddyByteCodeStrategy
import com.runninglane.dto.buddy.bytecode.bytebuddy.ByteBuddyWrapper
import net.bytebuddy.dynamic.DynamicType
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

class ByteCodeStrategyTest {
    abstract class TestDto

    annotation class TestAnnotation(val value: String)

    class TestByteBuddyWrapper : ByteBuddyWrapper() {
        override fun createDynamicType(
            baseClass: Class<*>,
            typeParams: List<Class<*>>?,
            packageName: String,
            className: String
        ): DynamicType.Builder<*> {
            return super.createDynamicType(baseClass, typeParams, packageName, className)
                .annotateType(TestAnnotation("test123"))
        }
    }

    @Test
    fun testAddSecondaryConstructor() {
        val dtoBuddy = DtoBuddy(ByteBuddyByteCodeStrategy(TestByteBuddyWrapper()))
        val concreteClass = dtoBuddy.implement(TestDto::class.java)
        val annotation = concreteClass.getAnnotation(TestAnnotation::class.java)
        assertNotNull(annotation)
        assertEquals("test123", annotation.value)
    }
}